File: src/components/home/DeliveryAddressDialog.js
import { useState } from "react";
import { useTranslation } from "react-i18next";
import { useDispatch } from "react-redux";
import Dialog from "@mui/material/Dialog";
import DialogTitle from "@mui/material/DialogTitle";
import DialogContent from "@mui/material/DialogContent";
import DialogActions from "@mui/material/DialogActions";
import Button from "@mui/material/Button";
import Box from "@mui/material/Box";
import Typography from "@mui/material/Typography";
import MyLocationRoundedIcon from "@mui/icons-material/MyLocationRounded";
import AddressAutocompleteField from "../maps/AddressAutocompleteField";
import { setDeliveryLocation } from "../../redux/slices/locationSlice";

/**
 * Opened from the home screen's AddressBar - lets the user either pick a
 * real Google Places suggestion (auto pick, geocoded) or use the device's
 * actual GPS position, reverse-geocoded to a short label via
 * google.maps.Geocoder. Both paths dispatch the same setDeliveryLocation
 * action, persisted via redux-persist so it survives a reload.
 */
export default function DeliveryAddressDialog({ open, onClose }) {
  const { t } = useTranslation();
  const dispatch = useDispatch();
  const [address, setAddress] = useState("");
  const [locating, setLocating] = useState(false);
  const [error, setError] = useState("");

  const handlePlaceSelected = ({ address: picked, lat, lng }) => {
    dispatch(setDeliveryLocation({ address: picked, lat, lng }));
    setAddress("");
    onClose();
  };

  const applyDetectedPosition = (lat, lng) => {
    if (window.google?.maps) {
      new window.google.maps.Geocoder().geocode({ location: { lat, lng } }, (results, status) => {
        setLocating(false);
        const label =
          status === "OK" && results?.[0]
            ? results[0].formatted_address
            : `${lat.toFixed(4)}, ${lng.toFixed(4)}`;
        dispatch(setDeliveryLocation({ address: label, lat, lng }));
        onClose();
      });
      return;
    }
    setLocating(false);
    dispatch(setDeliveryLocation({ address: `${lat.toFixed(4)}, ${lng.toFixed(4)}`, lat, lng }));
    onClose();
  };

  const handleUseMyLocation = () => {
    if (!navigator.geolocation) {
      setError(t("common.locationError"));
      return;
    }
    setError("");
    setLocating(true);
    navigator.geolocation.getCurrentPosition(
      (pos) => applyDetectedPosition(pos.coords.latitude, pos.coords.longitude),
      () => {
        setLocating(false);
        setError(t("common.locationError"));
      }
    );
  };

  return (
    <Dialog open={open} onClose={onClose} fullWidth maxWidth="xs">
      <DialogTitle>{t("common.setDeliveryAddress")}</DialogTitle>
      <DialogContent>
        <Box sx={{ display: "flex", flexDirection: "column", gap: 2, pt: 0.5 }}>
          <AddressAutocompleteField
            label={t("common.deliveryAddress")}
            fullWidth
            value={address}
            onTextChange={setAddress}
            onPlaceSelected={handlePlaceSelected}
          />
          <Button
            variant="outlined"
            startIcon={<MyLocationRoundedIcon />}
            onClick={handleUseMyLocation}
            disabled={locating}
          >
            {locating ? t("common.loading") : t("common.useMyLocation")}
          </Button>
          {error && (
            <Typography variant="caption" sx={{ color: "error.main" }}>
              {error}
            </Typography>
          )}
        </Box>
      </DialogContent>
      <DialogActions>
        <Button onClick={onClose}>{t("common.cancel")}</Button>
      </DialogActions>
    </Dialog>
  );
}

File: mobile/lib/widgets/delivery_address_dialog.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';

import '../core/geo.dart';
import '../core/places.dart';
import '../l10n/app_localizations.dart';
import '../providers/location_provider.dart';
import 'address_autocomplete_field.dart';

/// Opened from the home screen's AddressBar - mirrors
/// src/components/home/DeliveryAddressDialog.js: pick a real Google Places
/// suggestion (auto pick, geocoded), or use the device's actual GPS
/// position reverse-geocoded to a short label. Both paths persist via
/// LocationProvider.
class DeliveryAddressDialog extends StatefulWidget {
  const DeliveryAddressDialog({super.key});

  @override
  State<DeliveryAddressDialog> createState() => _DeliveryAddressDialogState();
}

class _DeliveryAddressDialogState extends State<DeliveryAddressDialog> {
  final _controller = TextEditingController();
  bool _locating = false;
  String? _error;

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  void _applyLocation(String address, double lat, double lng) {
    context.read<LocationProvider>().setLocation(address: address, lat: lat, lng: lng);
    Navigator.of(context).pop();
  }

  Future<void> _useMyLocation() async {
    setState(() {
      _error = null;
      _locating = true;
    });
    final position = await getCurrentLatLng();
    if (!mounted) return;
    if (position == null) {
      setState(() {
        _locating = false;
        _error = context.tr('common.locationError');
      });
      return;
    }
    final (lat, lng) = position;
    final address = await reverseGeocode(lat, lng) ?? '${lat.toStringAsFixed(4)}, ${lng.toStringAsFixed(4)}';
    if (!mounted) return;
    setState(() => _locating = false);
    _applyLocation(address, lat, lng);
  }

  @override
  Widget build(BuildContext context) {
    return AlertDialog(
      title: Text(context.t('common.setDeliveryAddress')),
      content: SingleChildScrollView(
        child: Column(
          mainAxisSize: MainAxisSize.min,
          crossAxisAlignment: CrossAxisAlignment.stretch,
          children: [
            AddressAutocompleteField(
              controller: _controller,
              label: context.t('common.deliveryAddress'),
              onPlaceSelected: ({required address, required lat, required lng}) =>
                  _applyLocation(address, lat, lng),
            ),
            const SizedBox(height: 12),
            OutlinedButton.icon(
              onPressed: _locating ? null : _useMyLocation,
              icon: const Icon(Icons.my_location_rounded),
              label: Text(_locating ? context.t('common.loading') : context.t('common.useMyLocation')),
            ),
            if (_error != null) ...[
              const SizedBox(height: 8),
              Text(_error!, style: const TextStyle(color: Colors.red, fontSize: 12)),
            ],
          ],
        ),
      ),
      actions: [
        TextButton(onPressed: () => Navigator.of(context).pop(), child: Text(context.t('common.cancel'))),
      ],
    );
  }
}
